src/Common/src/Interop/Unix/libcurl/Interop.SafeCurlMultiHandle.cs

```diff
@@ -0,0 +1,237 @@
+                            }
+                            numFds--;
+                        }
+
+                        // Now check for events on the remaining fds
+                        for (int i = 1; i < count && numFds > 0; i++)
+                        {
+                            if (fds[i].revents == 0)
+                            {
+                                continue;
+                            }
+                            readyFds.Add(fds[i]);
+                            numFds--;
+                        }
+                    }
+                }
+            }
+
+            internal void SignalFdSetChange(int fd, bool isRemove)
+            {
+                Debug.Assert(Monitor.IsEntered(this));
+                bool changed = isRemove ? _fdSet.Remove(fd) : _fdSet.Add(fd);
+                if (!changed)
+                {
+                    return;
+                }
+
+                unsafe
+                {
+                    // Write to special fd
+                    byte* dummyBytes = stackalloc byte[1];
+                    if ((int)libc.write(_specialFds[libc.WriteEndOfPipe], dummyBytes, (size_t)1) <= 0)
+                    {
+                        // TODO: How to handle errors?
+                        throw new InvalidOperationException("Cannot write data: " + Marshal.GetLastWin32Error());
+                    }
+                }
+            }
+
+            protected override void Dispose(bool disposing)
+            {
+                if (disposing)
+                {
+                    if (null != _timer)
+                    {
+                        _timer.Dispose();
+                    }
+                }
+                base.Dispose(disposing);
+            }
+
+            protected override bool ReleaseHandle()
+            {
+                Debug.Assert(0 == _fdSet.Count);
+                Debug.Assert(0 == _requestCount);
+                Debug.Assert(_pollCancelled);
+
+                Interop.Sys.Close(_specialFds[libc.ReadEndOfPipe]);
+                Interop.Sys.Close(_specialFds[libc.WriteEndOfPipe]);
+                libcurl.curl_multi_cleanup(this.handle);
+
+                return true;
+            }
+
+            private int ReadSpecialFd(PollFlags revents)
+            {
+                PollFlags badEvents = PollFlags.POLLERR | PollFlags.POLLHUP | PollFlags.POLLNVAL;
+                if ((revents & badEvents) != 0)
+                {
+                    return -1;
+                }
+                Debug.Assert((revents & PollFlags.POLLIN) != 0);
+                int pipeReadFd = _specialFds[libc.ReadEndOfPipe];
+                int bytesRead = 0;
+                unsafe
+                {
+                    do
+                    {
+                        // Read available data from the pipe
+                        int bufferLength = 1024;
```

Comment:
This is part of a non-semantic partitioning of Interop.libcurl.cs into separated files; it's not net new code.  (Though I do agree with you).